nix-parser/grammar.abnf

```diff
@@ -76,12 +77,27 @@ expr =
     with expr semicolon ; backtrack
   / assert expr semicolon ; backtrack
   / let set-bindings in expr ; backtrack
-  ; function
-  / left-curly-bracket formal-args right-curly-bracket [ at ident ] colon expr ; backtrack
-  / ident at left-curly-bracket formal-args right-curly-bracket colon expr ; backtrack
+  / fn-decl
+
+  ; expression with operators at lowest precedences
   / logic-imply-expr
   / sum-expr
 
+; function
+; the following non-terminal is an alias in the sense that
+; it should be "inlined" in all references to it, and thus
+; backtracking is "inlined" in the parsing logic of the
+; upper level non-terminals, instead of producing fatal
+; parser error
+fn-decl =
```

Comment:
@ebkalderon I just realize that we have never define the grammar rules for functions properly :rofl: